src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,165 @@
+-- vector of runs and vector of keys respectively.
+--
+-- Note: the result vector should be ephemeral, so do not retain it.
 --
--- Note: results are grouped by key instead of file descriptor, because this
--- means that results for a single key are close together.
-prepLookups :: [Run fd] -> [SerialisedKey] -> [(SerialisedKey, (fd, PageSpan))]
-prepLookups runs ks =
-    [ (k, (fd, pspan))
-    | k <- ks
-    , r@(fd,_,_) <- runs
-    , pspan <- toList (prepLookup r k)
-    ]
-
-prepLookup :: Run fd -> SerialisedKey -> Maybe PageSpan
-prepLookup (_fd, b, fpix) k
-  | Bloom.elem k b = Index.toPageSpan $ Index.search k fpix
-  | otherwise      = Nothing
+-- TODO: we consider it likely that we could implement a optimised, batched
+-- version of bloom filter queries, which would largely replace this function.
+bloomQueries :: V.Vector (RunLookupView h) -> V.Vector SerialisedKey -> VU.Vector (RunIx, KeyIx)
+bloomQueries rs ks = VU.create $ do
+      -- The result vector can be of variable length, so we use a generous
+      -- estimate here, and we grow the vector if needed. TODO: tune the
+      -- starting estimate based on the expected true- and false-positives.
+      res <- VUM.unsafeNew (2 * ksN)
+      resixVar <- newPrimVar 0
+      res' <- go res resixVar
+      resix <- readPrimVar resixVar
+      -- The result vector is not guaranteed to be full, so we slice it. This
+      -- means it might retain memory that is not used by the vector.
+      pure $ VUM.slice 0 resix res'
+  where
+    !rsN = V.length rs
+    !ksN = V.length ks
+
+    go :: forall s.
+         VUM.MVector s (RunIx, KeyIx)
+      -> PrimVar s ResIx
+      -> ST s (VUM.MVector s (RunIx, KeyIx))
+    go res0 resixVar = loop1 res0 0
+      where
+        -- Loop over all run indexes
+        loop1 ::
+             VUM.MVector s (RunIx, KeyIx)
+          -> RunIx
+          -> ST s (VUM.MVector s (RunIx, KeyIx))
+        loop1 !res1 !rix
+          | rix == rsN = pure res1
+          | otherwise
+          = do
+              res1' <- loop2 res1 0 (rlvBloom $ rs `V.unsafeIndex` rix)
+              loop1 res1' (rix+1)
+          where
+            -- Loop over all key indexes
+            loop2 ::
+                 VUM.MVector s (RunIx, KeyIx)
+              -> KeyIx
+              -> Bloom SerialisedKey
+              -> ST s (VUM.MVector s (RunIx, KeyIx))
+            loop2 !res2 !kix b
+              | kix == ksN = pure res2
+              | let !k = ks `V.unsafeIndex` kix
+              , Bloom.elem k b = do
+                  resix <- readPrimVar resixVar
+                  -- Grows the vector if we've reached the end. TODO: tune how
+                  -- much much we grow the vector each time based on the
+                  -- expected true- and false-positives.
+                  res2' <- if resix == VUM.length res2
+                            then VUM.unsafeGrow res2 ksN
+                            else pure res2
```

Comment:
The initial size seems reasonable and we can tune the numbers, but generally my intuition would be to grow the vector exponentially instead of by a fixed size, so we limit the number of copies in extreme cases.